node/src/components/linear_chain.rs

```diff
@@ -203,3 +210,445 @@ where
         }
     }
 }
+
+/// A trait to allow an `EffectBuilder` to be replaced in testing by a mock.
+#[async_trait]
+pub(crate) trait EraValidatorsGetter: Copy {
+    async fn get_switch_block_header_at_era_id_from_storage(
+        self,
+        era_id: EraId,
+    ) -> Option<BlockHeader>;
+
+    async fn get_era_validators_from_contract_runtime(
+        self,
+        request: EraValidatorsRequest,
+    ) -> Result<EraValidators, GetEraValidatorsError>;
+}
+
+#[async_trait]
+impl<REv> EraValidatorsGetter for EffectBuilder<REv>
+where
+    REv: From<StorageRequest> + From<ContractRuntimeRequest> + Send,
+{
+    async fn get_switch_block_header_at_era_id_from_storage(
+        self,
+        era_id: EraId,
+    ) -> Option<BlockHeader> {
+        self.get_switch_block_header_at_era_id_from_storage(era_id)
+            .await
```

Comment:
Are we sure this calls the right method - ie. the one implemented directly on `EffectBuilder`, and not itself, making it infinitely recursive? (Hopefully at least the tests would catch the infinite recursion if it happened, but I just want to make sure :sweat_smile: ). And same for the other method.

Reply:
Yes - the nctl tests would indeed have failed if there were a problem like this :)

It maybe helps that the `EraValidatorsGetter` can't be made into an object as it's not object safe, so we couldn't do e.g. `(self as EraValidatorsGetter).get_switch_block_header_at_era_id_from_storage`?